fix usage of Duration in Promise impl

rkuhn opened the pull request:
- correctly treat MinusInf and Undefined
- don't toMillis in the timeout message (could be MinusInf)
- also notice that Inf did not actually wait unbounded
- and further notice that tryAwait swallows InterruptedException instead
  of bailing out early => changed to do so and added throws annotation
- also removed some unused imports of Duration

rkuhn committed:
fix usage of Duration in Promise impl

- correctly treat MinusInf and Undefined
- don't toMillis in the timeout message (could be MinusInf)
- also notice that Inf did not actually wait unbounded
- and further notice that tryAwait swallows InterruptedException instead
  of bailing out early => changed to do so and added throws annotation
- also removed some unused imports of Duration

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/398/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1106/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/398/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1106/

jsuereth commented:
Review by?  I'll take it for now.

jsuereth commented:
LGTM.

jsuereth commented:
Also, I wish there a better way to test waiting unbounded....

jsuereth commented:
How bout a second review form @viktorklang @phaller or @heathermiller

viktorklang commented on src/library/scala/concurrent/Awaitable.scala:
```diff
@@ -18,13 +18,18 @@ trait Awaitable[+T] {
+   * @throws InterruptedException if the wait call was interrupted
    */
   @throws(classOf[TimeoutException])
+  @throws(classOf[InterruptedException])
```
Oh, is that you you add multiple throws clauses? Has this been verified?

rkuhn replied:
yes, I verified the byte code

viktorklang commented on src/library/scala/concurrent/Awaitable.scala:
```diff
@@ -18,13 +18,18 @@ trait Awaitable[+T] {
    * Throws exceptions if it cannot produce a T within the specified time.
    * This method should not be called directly.
+   *
+   * @throws InterruptedException if the wait call was interrupted
```
Could also throw something else (failure to produce a result)

rkuhn replied:
Yes, I was taking that as a given. But I agree that the comment above could also be improved.

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -65,29 +65,44 @@ private[concurrent] object Promise {
       @tailrec
       def awaitUnsafe(waitTimeNanos: Long): Boolean = {
-        if (value.isEmpty && waitTimeNanos > 0) {
+        if (!isCompleted && waitTimeNanos > 0) {
```
nice

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -65,29 +65,44 @@ private[concurrent] object Promise {
-            case e: InterruptedException =>
+          synchronized {
+            if (!isCompleted) wait(ms, ns) // previously - this was a `while`, ending up in an infinite loop
           }
```
I'd make it a one-liner, but it's just a matter of taste

rkuhn replied:
Yeah, I left the comment intact, because I cannot relate to that “previous” thing. Might as well clean it up, I guess it was related to catching IE.

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -65,29 +65,44 @@ private[concurrent] object Promise {
+
+      if (atMost <= Duration.Zero)
+        isCompleted
+      else if (atMost eq Duration.Undefined)
```
I'd start with this check

rkuhn replied:
why does it matter?

viktorklang replied:
Because I think it's clearer to start with the most specific and then go less and less specific until "else" which is the least specific == catchall.

viktorklang replied:
Also I don't need to know whether undefined <= Zero or not

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -65,29 +65,44 @@ private[concurrent] object Promise {
     def ready(atMost: Duration)(implicit permit: CanAwait): this.type =
       if (isCompleted || tryAwait(atMost)) this
-      else throw new TimeoutException("Futures timed out after [" + atMost.toMillis + "] milliseconds")
+      else throw new TimeoutException("Futures timed out after [" + atMost + "]")
```
Great fix

viktorklang commented on src/library/scala/concurrent/package.scala:
```diff
@@ -74,8 +74,10 @@ package concurrent {
      * @param awaitable   the `Awaitable` on which `ready` is to be called
      * @param atMost      the maximum timeout for which to wait
      * @return            the result of `awaitable.ready` which is defined to be the awaitable itself.
+     * @throws InterruptedException if the wait call was interrupted
```
I'd clarify the other throws as well

viktorklang commented on test/files/jvm/scala-concurrent-tck.scala:
```diff
@@ -663,6 +672,29 @@ trait FutureProjections extends TestBase {
+      intercept[IllegalArgumentException] { Await.ready(f, Duration.Undefined) }
+      p.success(0)
+      Await.ready(f, Duration.Zero)
+      Await.ready(f, Duration(500, "ms"))
```
No test to verify when it's not completed yet?

rkuhn replied:
these tests are just for the Await-vs-Duration part of the API, the rest should be covered by the existing tests

jsuereth commented on test/files/jvm/scala-concurrent-tck.scala:
```diff
@@ -19,6 +20,14 @@ trait TestBase {
     sv.take(2000)
   }
+
+  def intercept[T <: Exception : ClassTag](code: => Unit): Unit =
```
Note:  it seems this method might be useful in a base partest class.   Mind if move this then?

rkuhn committed:
improve docs and Promise impl

- scaladoc the exceptions thrown by Await.* and Awaitable.*
- move intercept[Exception] into partest’s TestUtil object
- improve Promise.tryAwait implementation following Viktor’s comments
  and make use of Deadline to avoid calling System.nanoTime too often

rkuhn commented:
PLS REBUILD ALL

rkuhn commented:
@viktorklang @jsuereth Would you please review after my latest changes? Thanks.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/411/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1117/

jsuereth commented on src/partest/scala/tools/partest/TestUtil.scala:
```diff
@@ -29,8 +31,16 @@ trait TestUtil {
     assert(mult <= acceptableMultiple, "Performance difference too great: multiple = " + mult)
   }
+
+  def intercept[T <: Exception : ClassTag](code: => Unit): Unit =
```
Yay!  I need to go refactor my tests now :)

viktorklang replied:
Technically intercept should return what it intercepted so you can do things like:

intercept[IllegalArgumentException]{ block}.getMessage must be === "OH NO YOU DIDN'T!"

rkuhn replied:
I think Josh just volunteered for that change, when he’ll need it ;-)

jsuereth commented:
LGTM, G = Great not just good.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/411/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1117/

viktorklang commented:
Excellently engineered Roland!
